### PR TITLE
Add DASHA baseline

### PR DESCRIPTION
**Pull request by k3nfalt**
### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: baselines/dasha/dasha/conf/base.yaml
Comment (jafermarq) [inline baselines/dasha/dasha/conf/base.yaml:17]: Out of curiosity, why is this introduced? This disables the (in my view) desirable Hydra behaviour of automatically creating an output directory for your experiment (where log and config are dumped). By default is created in the top-level dasha directory (i.e. sitting next to the .toml) and includes a sub-directory structure with date/time when the experiment is run. 

This is the recommended place to store the experiment outputs. (you can retrieve this directory's path with `save_path = HydraConfig.get().runtime.output_dir`)

File: baselines/dasha/README.md
Comment (jafermarq) [inline baselines/dasha/README.md:118]: Could titles to the plots be added? (maybe also gridlines?)

File: baselines/dasha/pyproject.toml
Comment (jafermarq) [inline baselines/dasha/pyproject.toml:42]: Let's raise to the latest version of Flower. You can simply replace the lines above with:
```
flwr = { extras = ["simulation"], version = "1.5.0" }
```

File: baselines/dasha/README.md
Comment (jafermarq) [inline baselines/dasha/README.md:109]: maybe replace with `python -m dasha.main <...>` for consistency? (also the command below)

**Review by jafermarq — COMMENTED** [review]
Hi @k3nfalt,

I was able to re-produce your experiments (the long ones for CIFAR are still running) by following the instructions in your readme.
I've done an initial pass through your baseline and I left a few comments (mostly very minor) related to your README and where to save datasets and results. For the latter part, i think it's better if everything (including datasets and results and plots) are downloaded/outputted to the `DASHA` directory instead of somewhere else in the filesystem.

File: baselines/dasha/README.md
Comment (jafermarq) [inline baselines/dasha/README.md:12]:
```suggestion
****Paper:**** [openreview.net/forum?id=VA1YpcNr7ul](https://openreview.net/forum?id=VA1YpcNr7ul)
```

File: baselines/dasha/pyproject.toml
Comment (jafermarq) [inline baselines/dasha/pyproject.toml:56]: Since we created the baseline_template directory and its content we have made some changes to its `pyproject.toml`, most notably raising the mypy version. Could you update your `pyproject.toml` with the new content in https://github.com/adap/flower/blob/main/baselines/baseline_template/pyproject.toml?

File: baselines/dasha/README.md
Comment (jafermarq) [inline baselines/dasha/README.md:2]: The rendered readme shows an error. I think it's because the `title` here should be between double quotes (`"<DASHA: ....>"`

**Review by jafermarq — COMMENTED** [review]
Hi, I added some extra few comments. Also, could you rename the `docs` directory you use for the images in the readme as `_static`, this will make it easier to integrate parts of your readme into the Flower docs ?

File: baselines/dasha/README.md
Comment (k3nfalt) [reply baselines/dasha/README.md:54]: How do you know that CUDA 11.8 will be installed by poetry? Where is it said in baselines/dasha/pyproject.toml?

Comment (jafermarq) [reply baselines/dasha/README.md:55]: It ships w/ torch etc, right? (unless you specify cpu-only). What i meant to say with that comment is that there's no additional action to be taken (i.e. install cuda separately or have it pre-installed on the system) by somebody that wants to run your baseline.

Comment (jafermarq) [reply baselines/dasha/README.md:55]: But if happy if the comment stays there if you think its relevant

Comment (k3nfalt) [reply baselines/dasha/README.md:55]: "But if happy if the comment stays there if you think its relevant"
It is possible that I may have missed something, so I decided to ask an additional question for clarification.
"It ships w/ torch etc, right?"
I don't think so. CUDA is a Nvidia driver that is installed separately. Based on the version of a driver, one should choose the appropriate PyTorch installer. For instance, if I have CUDA 11.7 installed in my system, I'm not sure I can install Pytorch with CUDA 11.8.

Anyway, it is a minor thing.

Comment (jafermarq) [reply baselines/dasha/README.md:55]: the default way of installing pytorch (e.g. via `pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118`) does get all the CUDA (in this case 11.8) stuff you need. So you don't need to install it manually in your OS. The `nvidia-driver` is naturally required in advance.

Let's leave your comment there

Comment (k3nfalt) [reply baselines/dasha/README.md]: I've just rewrote the sentence to make it more clear.

File: baselines/dasha/README.md
Comment (k3nfalt) [reply baselines/dasha/README.md:2]: I think I fixed it. The problem was in other places.

File: baselines/dasha/dasha/conf/base.yaml
Comment (k3nfalt) [reply baselines/dasha/dasha/conf/base.yaml]: I'll return the expected behaviour.

> Out of curiosity, why is this introduced?

When I were developing and testing this project, I was using the repo in the read-only mode in a remote server. Locally, I could modify the repo, but remotely, where I executed the code, I could not do that. For instance, I could not produce outputs inside the repo. So I decided to save everything separate directories.

File: baselines/dasha/README.md
Comment (k3nfalt) [reply baselines/dasha/README.md]: By default, they will be generated in the Hydra's output default path (with data and time) in the dasha directory.

**Comment by k3nfalt** [comment]
@jafermarq, hi!

I've responded and fixed all comments. Please let me know if I missed something.

File: baselines/dasha/README.md
Comment (jafermarq) [inline baselines/dasha/README.md:57]: We very recently changed the minimum version of python for baselines to 3.10. Could you reflect that in your comment above?
Also, i'm thinking on adding a complete description of how to setup the env using `pyenv`. This could be done by adding the following likes before your `poetry install`:

```
# Set Python 3.10
pyenv local 3.10.6
# Tell poetry to use python 3.10
poetry env use 3.10.6
# Now install your environment
```

See my other edit suggestion for the `pyproject.toml` to update to py3.10

File: baselines/dasha/pyproject.toml
Comment (jafermarq) [inline baselines/dasha/pyproject.toml:55]:
```suggestion
mypy = "==1.4.1"
```

**Review by jafermarq — COMMENTED** [review]
Hi @k3nfalt,

Thanks for adding all those small changes. I have done another pass and made some further comments. Also:
* Could you add grid lines to the plots? (I think i forgot to mention this in my previous review).
* I ran the formatting and tests scripts and both finished fine! cool.
* I'm not quite sure why the edit to the `fedprox` readme is showing up as part of your baseline. Could you check this?

**Comment by k3nfalt** [comment]
@jafermarq, I've just fixed the additional issues (grid lines, fed prox). I also added the support of py10 and tested it that it works.

File: baselines/dasha/README.md
Comment (jafermarq) [inline baselines/dasha/README.md:120]:
```suggestion

python -m dasha.main method=stochastic_marina method.strategy.step_size=0.01 num_rounds=10000 compressor.number_of_coordinates=2000000 model=resnet_18_with_logistic_loss method.client.strict_load=false dataset=cifar10 method.client.device=cuda method.client.evaluate_accuracy=true local_address=localhost:8002
```

Comment (jafermarq) [reply baselines/dasha/README.md:120]: just adding a bit of space between both commands

File: baselines/dasha/README.md
Comment (jafermarq) [inline baselines/dasha/README.md:105]: Do you want to keep this here? (there is another a bit below in the readme too)

File: baselines/dasha/dasha/client.py
Comment (jafermarq) [inline baselines/dasha/dasha/client.py:1]: Many methods are set as `_method_name` making them not publich (and therefore not required by the formatting/tests checks to include some minimal comment of what it does). Could you please add some comments ? The idea is if someone else wants to run/understand your code they should know a bit what's going on, for instance, in `_stochastic_compression_step()`, what's `mega_batch_size`, etc. If you think it's useful to add a ref to subsections/equations in the paper, then that's cool too.

File: baselines/dasha/pyproject.toml
Comment (jafermarq) [inline baselines/dasha/pyproject.toml:40]:
```suggestion
python = ">=3.10.0, <3.11.0"  # don't change this
```

**Review by jafermarq — COMMENTED** [review]
Hi @k3nfalt ,

This looks good. I only left a few comments about the README.md. Also, could you please add a bit more documentation to at least the `client.py` and `strategy.py`. I think it will be useful for people out there that are curious to dive into DASHA. 

After that, we can start the process of merging it with Flower's `main` branch. All formatting/test checks pass so we are good in that front. I was able to reproduce your experiments.

File: baselines/dasha/dasha/client.py
Comment (k3nfalt) [reply baselines/dasha/dasha/client.py:1]: Ok, done!

Comment (jafermarq) [reply baselines/dasha/dasha/client.py:1]: thanks!!

**Comment by jafermarq** [comment]
Hi @k3nfalt, I'm wondering if we should keep the `test_dasha_baseline.py` test, it takes quite some time to run (too much for our CI I think) and it is seems to not guarantee to pass (maybe because only two clients are involved?). The assert was triggered. On my machine i was able to run it just fine several times, also when lowering the rounds to 100. 

There is value in keeping these tests around. maybe we can rename them or so they are not executed by you `run_tests.sh`? How do you want to proceed with this?

**Comment by k3nfalt** [comment]
Hi @jafermarq, I propose to make them optional. By default, they will be turned off, but anyone can run them with an appropriate flag. I will add this information to README.

**Comment by k3nfalt** [comment]
I accidentally changed the wrong file "femnist/dataset/dataset_test.py". I reversed the change.

**Comment by k3nfalt** [comment]
Let's run the tests again?

**Comment by jafermarq** [comment]
> Let's run the tests again?

Have you re-run the formatting scripts just in case?

**Comment by k3nfalt** [comment]
I ran "sh ./dev/test-baseline.sh dasha" and it worked.

File: baselines/dasha/pyproject.toml
Comment (jafermarq) [inline baselines/dasha/pyproject.toml:10]: Update this with your details if you'd like. Also, the `description` field above could just say "Dasha" or maybe the whole paper title. Up to you!

**Review by jafermarq — COMMENTED** [review]
Hi @k3nfalt,

Just a small comment for the `pyproject.toml`. I think this is ready to be merge. Update your branch with the content of `main` when you have a sec.

**Comment by jafermarq** [comment]
Hi @k3nfalt,

Your `DASHA` baseline can be merged once you do the small edit to `pyproject.toml` proposed above and you merge the latest Flower `main` into your branch. Ping me once this is ready!

**Comment by k3nfalt** [comment]
@jafermarq, I've finished the changes.

**Review by jafermarq — APPROVED** [review]
Look great! @k3nfalt

**Comment by jafermarq** [comment]
Hi @k3nfalt , your baseline will be automatically merged once you merge (again, sorry i wasn't quick enough last time) `main` into your fork/branch.

**Comment by k3nfalt** [comment]
@jafermarq, I've just updated the branch

**Pull request merged**
